File: backend/app/core/form_rules.py
from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from datetime import date, datetime
from typing import Any

from pydantic.alias_generators import to_camel


@dataclass
class RuleSpec:
    name: str  # camelCase keyword name, e.g. "mustMatch"
    params: dict[str, Any]  # camelCase field names for x-rules emission
    validate: Callable[[Any], None]


class FormRuleRegistry:
    _rules: dict[str, Callable[..., RuleSpec]] = {}

    @classmethod
    def register(cls, name: str, factory: Callable[..., RuleSpec]) -> None:
        if name in cls._rules:
            raise ValueError(f"rule {name!r} already registered")
        cls._rules[name] = factory

    @classmethod
    def is_registered(cls, name: str) -> bool:
        return name in cls._rules


def must_match(*, a: str, b: str) -> RuleSpec:
    def _check(instance: Any) -> None:
        va = getattr(instance, a)
        vb = getattr(instance, b)
        if va != vb:
            raise ValueError(f"{a} must equal {b}")

    return RuleSpec(
        name="mustMatch",
        params={"a": to_camel(a), "b": to_camel(b)},
        validate=_check,
    )


def date_order(*, start: str, end: str) -> RuleSpec:
    def _check(instance: Any) -> None:
        s = getattr(instance, start, None)
        e = getattr(instance, end, None)
        if s is None or e is None:
            return
        if not isinstance(s, (date, datetime)) or not isinstance(e, (date, datetime)):
            raise ValueError(f"{start} and {end} must be dates")
        if e <= s:
            raise ValueError(f"{end} must be after {start}")

    return RuleSpec(
        name="dateOrder",
        params={"start": to_camel(start), "end": to_camel(end)},
        validate=_check,
    )


def password_policy(field: str) -> RuleSpec:
    camel_field = to_camel(field)

    def _check(instance: Any) -> None:
        value = getattr(instance, field, "")
        if not isinstance(value, str):
            return
        if len(value) < 10:
            raise ValueError("Password must be at least 10 characters")
        has_letter = any(c.isalpha() for c in value)
        has_digit = any(c.isdigit() for c in value)
        if not has_letter or not has_digit:
            raise ValueError("Password must contain at least 1 letter and 1 digit")

    return RuleSpec(
        name="passwordPolicy",
        params={"field": camel_field},
        validate=_check,
    )


FormRuleRegistry.register("mustMatch", must_match)
FormRuleRegistry.register("dateOrder", date_order)
FormRuleRegistry.register("passwordPolicy", password_policy)

File: frontend/src/lib/form-rules.ts
import type Ajv from "ajv";

interface MustMatchParams {
  a: string;
  b: string;
}

interface DateOrderParams {
  start: string;
  end: string;
}

export function registerRuleKeywords(ajv: Ajv): void {
  ajv.addKeyword({
    keyword: "mustMatch",
    type: "object",
    errors: true,
    validate: function mustMatch(params: MustMatchParams, data: Record<string, unknown>) {
      const ok = data[params.a] === data[params.b];
      if (!ok) {
        // @ts-expect-error — Ajv attaches errors on the function
        mustMatch.errors = [
          { keyword: "mustMatch", message: `${params.a} must equal ${params.b}`, params },
        ];
      }
      return ok;
    },
  });

  ajv.addKeyword({
    keyword: "dateOrder",
    type: "object",
    errors: true,
    validate: function dateOrder(params: DateOrderParams, data: Record<string, unknown>) {
      const s = data[params.start];
      const e = data[params.end];
      if (s == null || e == null) return true;
      if (typeof s !== "string" || typeof e !== "string") return false;
      const ok = new Date(e) > new Date(s);
      if (!ok) {
        // @ts-expect-error — Ajv attaches errors on the function
        dateOrder.errors = [
          { keyword: "dateOrder", message: `${params.end} must be after ${params.start}`, params },
        ];
      }
      return ok;
    },
  });
}
